drivers/platform/xilinx/spi.c

```diff
@@ -54,11 +54,98 @@
 #include "error.h"
 #include "spi.h"
 #include "spi_extra.h"
+#include "spi_engine.h"
 
 /******************************************************************************/
 /************************ Functions Definitions *******************************/
 /******************************************************************************/
 
+int32_t spi_init_pl(struct xil_spi_desc *xdesc,
+		    struct xil_spi_init_param *xinit)
+{
+#ifdef XSPI_H
+	int32_t ret;
+
+	desc->instance = (XSpi*)malloc(sizeof(XSpi));
+	if(!desc->instance)
+		goto pl_error;
+
+	xdesc->config = XSpi_LookupConfig(xinit->device_id);
+	if(xdesc->config == NULL)
+		goto pl_error;
+
+	ret = XSpi_CfgInitialize(xdesc->instance,
+				 xdesc->config,
+				 ((XSpi_Config*)xdesc->config)
+				 ->BaseAddress);
+	if(ret != SUCCESS)
+		goto pl_error;
+
+	ret = XSpi_Initialize(xdesc->instance, xinit->device_id);
+	if (ret != 0)
+		goto pl_error;
+
+	ret = XSpi_SetOptions(xdesc->instance,
+			      XSP_MASTER_OPTION |
+			      ((sdesc->mode & SPI_CPOL) ?
+			       XSP_CLK_ACTIVE_LOW_OPTION : 0) |
+			      ((sdesc->mode & SPI_CPHA) ?
+			       XSP_CLK_PHASE_1_OPTION : 0));
+	if (ret != 0)
+		goto pl_error;
+
+	ret = XSpi_Start(xdesc->instance);
+	if (ret != 0)
+		goto pl_error;
+
+	XSpi_IntrGlobalDisable((XSpi *)(xdesc->instance));
+
+	return SUCCESS;
+
+pl_error:
+	free(xdesc->instance);
+#endif
+	free(xdesc);
+
+	return FAILURE;
+}
+
+int32_t spi_init_ps(struct xil_spi_desc *xdesc,
+		    struct xil_spi_init_param *xinit)
+{
+#ifdef XSPIPS_H
+	int32_t ret;
+
+	xdesc->instance = (XSpiPs*)malloc(sizeof(XSpiPs));
+	if(!xdesc->instance)
+		goto ps_error;
+
+	xdesc->config = XSpiPs_LookupConfig(xinit->device_id);
+	if(xdesc->config == NULL)
+		goto ps_error;
+
+	ret = XSpiPs_CfgInitialize(xdesc->instance,
+				   xdesc->config,
+				   ((XSpiPs_Config*)xdesc->config)
+				   ->BaseAddress);
+	if(ret != SUCCESS)
+		goto ps_error;
+
+	ret = XSpiPs_SetClkPrescaler(xdesc->instance,
+				     XSPIPS_CLK_PRESCALE_64);
+	if(ret != SUCCESS)
+		goto ps_error;
+
+	return SUCCESS;
+
+ps_error:
+	free(xdesc->instance);
+#endif
+	free(xdesc);
```

Comment:
this free must not be done here. It is done in the outside function. Same for init_pl

Reply:
I believe you are suggesting that the user should check the init and call the remove function himself in case of error. In this way, again, we are forcing the user to call additional functions.